google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/reframing/ReframingResponseObserver.java

```diff
@@ -226,155 +231,150 @@ protected void onErrorImpl(Throwable t) {
-            }
-            closed = true;
-            closeError = error;
-            action = DeliveryAction.CLOSE;
-          }
-          // demand has been fulfilled, do nothing
-          else if (numPending == 0) {
-            action = DeliveryAction.FULFILLED;
-          } else {
-            throw new IllegalStateException("ReframingResponseObserver is in an unexpected state");
+      // Process the upstream message if one exists.
+      pollUpstream();
+
+      // Eagerly notify of onComplete/onError disregarding demand.
+      // NOTE: this will purposely leave wip set to avoid further deliveries.
+      if (maybeFinish()) {
+        return;
+      }
+
+      // Deliver as many messages as possible
+      int demandSnapshot = numRequested.get();
+      int delivered = 0;
+
+      while (delivered < demandSnapshot) {
+        // Deliver a message if we can
+        if (reframer.hasFullFrame()) {
+          delivered++;
+          outerResponseObserver.onResponse(reframer.pop());
+        } else {
+          // Otherwise request more from upstream (if we haven't done so already)
+          if (!awaitingInner) {
+            awaitingInner = true;
+            innerController.request(1);
           }
+          break;
         }
 
-        if (action == DeliveryAction.DELIVER) {
-          outerResponseObserver.onResponse(result);
+        if (maybeFinish()) {
+          return;
         }
-      } while (action == DeliveryAction.DELIVER);
+      }
 
-      switch (action) {
-        case REQUEST_MORE:
-          innerController.request(1);
-          break;
-        case CLOSE:
-          if (closeError != null) {
-            outerResponseObserver.onError(closeError);
-          } else {
-            outerResponseObserver.onComplete();
-          }
-          break;
+      // update the counter in bulk
+      if (delivered != 0) {
+        numRequested.addAndGet(-delivered);
       }
+    } while (lock.decrementAndGet() != 0);
+  }
 
-      // exit only if there were no missed delivery requests
-      synchronized (lock) {
-        if (missed) {
-          missed = false;
-          continue;
-        }
+  /**
+   * Checks if the awaited upstream response is available. If it is, then feed it to the {@link
+   * Reframer} and update the {@link #awaitingInner} flag. Upon exit, if awaitingInner is not set,
+   * then done is guaranteed to reflect the current status of the upstream.
+   */
+  private void pollUpstream() {
+    if (!awaitingInner) {
+      return;
+    }
 
-        inDelivery = false;
-        return;
+    boolean localDone = this.done;
+
+    // Try to move the new item into the reframer
+    InnerT newUpstreamItem = newItem.getAndSet(null);
+    if (newUpstreamItem != null) {
+      reframer.push(newUpstreamItem);
+      // and reset the awaiting flag, if the item arrived or upstream closed
+      awaitingInner = false;
+    } else if (localDone) {
+      awaitingInner = false;
+    }
```

Comment:
Isn't this just

```java
if (newUpstreamItem != null) {
  reframer.push(newUpstreamItem);
}
awaitingInner = false;
```

Reply:
I don't think so. Though you might be seeing something I'm not. The general order of operations in delviery is:

1. check if we are awaitingInner, which we aren't
2. notice that delivered < demand
3. ask the reframer for a full frame, which it would not have
4. check that we aren't awaitingInner, & request(1) from inner controller
5. go back to the top of the outer loop
6. here we have a race condition: grpc might or might have called onResponse and/or onComplete
  a. if it didn't, then awaitingInner is true and should remain so
  b. if it did, then we have a newupstreamItem and we should reset awaitingInner
  c. if grpc called onComplete, then awaitinginner should be reset as well

automatically setting awaitingInner=false would neglect 6a

Reply:
Right sorry. Wasn't thinking.